File: socialgo/src/main/java/com/fungo/socialgo/platform/wechat/pay/WXPay.java
package com.fungo.socialgo.platform.wechat.pay;

import android.content.Context;
import android.text.TextUtils;

import com.fungo.socialgo.listener.OnPayListener;
import com.tencent.mm.sdk.constants.Build;
import com.tencent.mm.sdk.modelpay.PayReq;
import com.tencent.mm.sdk.openapi.IWXAPI;
import com.tencent.mm.sdk.openapi.WXAPIFactory;

import org.json.JSONException;
import org.json.JSONObject;

/**
 * 微信支付
 * Created by tsy on 16/6/1.
 */
public class WXPay {

    private static WXPay mWXPay;
    private IWXAPI mWXApi;
    private String mPayParam;
    private OnPayListener mListener;

    public static final int NO_OR_LOW_WX = 1;   //未安装微信或微信版本过低
    public static final int ERROR_PAY_PARAM = 2;  //支付参数错误
    public static final int ERROR_PAY = 3;  //支付失败

    public WXPay(Context context, String wx_appid) {
        mWXApi = WXAPIFactory.createWXAPI(context, null);
        mWXApi.registerApp(wx_appid);
    }

    public static void init(Context context, String wx_appid) {
        if(mWXPay == null) {
            mWXPay = new WXPay(context, wx_appid);
        }
    }
    public static WXPay getInstance(){
        return mWXPay;
    }

    public IWXAPI getWXApi() {
        return mWXApi;
    }
    /**
     * 发起微信支付
     */
    public void doPay(String pay_param, OnPayListener listener) {
        mPayParam = pay_param;
        mListener = listener;

        if(!check()) {
            if(mListener != null) {
                mListener.onError(NO_OR_LOW_WX);
            }
            return;
        }

        JSONObject param = null;
        try {
            param = new JSONObject(mPayParam);
        } catch (JSONException e) {
            e.printStackTrace();
            if(mListener != null) {
                mListener.onError(ERROR_PAY_PARAM);
            }
            return;
        }
        if(TextUtils.isEmpty(param.optString("appid")) || TextUtils.isEmpty(param.optString("partnerid"))
                || TextUtils.isEmpty(param.optString("prepayid")) || TextUtils.isEmpty(param.optString("package")) ||
                TextUtils.isEmpty(param.optString("noncestr")) || TextUtils.isEmpty(param.optString("timestamp")) ||
                TextUtils.isEmpty(param.optString("sign"))) {
            if(mListener != null) {
                mListener.onError(ERROR_PAY_PARAM);
            }
            return;
        }

        PayReq req = new PayReq();
        req.appId = param.optString("appid");
        req.partnerId = param.optString("partnerid");
        req.prepayId = param.optString("prepayid");
        req.packageValue = param.optString("package");
        req.nonceStr = param.optString("noncestr");
        req.timeStamp = param.optString("timestamp");
        req.sign = param.optString("sign");

        mWXApi.sendReq(req);
    }

    //支付回调响应
    public void onResp(int error_code) {
        if(mListener == null) {
            return;
        }

        if(error_code == 0) {   //成功
            mListener.onSuccess();
        } else if(error_code == -1) {   //错误
            mListener.onError(ERROR_PAY);
        } else if(error_code == -2) {   //取消
            mListener.onCancel();
        }

        mListener = null;
    }

    //检测是否支持微信支付
    private boolean check() {
        return mWXApi.isWXAppInstalled() && mWXApi.getWXAppSupportAPI() >= Build.PAY_SUPPORTED_SDK_INT;
    }
}

File: socialgo/src/main/java/com/fungo/socialgo/platform/wechat/pay/WXPay.kt
package com.fungo.socialgo.platform.wechat.pay

import android.text.TextUtils

import com.fungo.socialgo.exception.SocialError
import com.fungo.socialgo.listener.OnPayListener
import com.tencent.mm.opensdk.constants.Build
import com.tencent.mm.opensdk.modelpay.PayReq
import com.tencent.mm.opensdk.openapi.IWXAPI

import org.json.JSONException
import org.json.JSONObject

/**
 * 微信支付
 */
class WXPay(private val wxApi: IWXAPI) {

    private var mPayParam: String? = null
    private var mListener: OnPayListener? = null

    /**
     * 发起微信支付
     */
    fun doPay(param: String, listener: OnPayListener?) {
        mPayParam = param
        mListener = listener

        if (!check()) {
            mListener?.onError(SocialError(SocialError.CODE_NOT_INSTALL))
            return
        }

        val json: JSONObject
        try {
            json = JSONObject(mPayParam)
        } catch (e: JSONException) {
            e.printStackTrace()
            mListener?.onError(SocialError(SocialError.CODE_PAY_PARAM_ERROR))
            return
        }

        if (TextUtils.isEmpty(json.optString("appid")) || TextUtils.isEmpty(json.optString("partnerid"))
                || TextUtils.isEmpty(json.optString("prepayid")) || TextUtils.isEmpty(json.optString("package")) ||
                TextUtils.isEmpty(json.optString("noncestr")) || TextUtils.isEmpty(json.optString("timestamp")) ||
                TextUtils.isEmpty(json.optString("sign"))) {
            mListener?.onError(SocialError(SocialError.CODE_PAY_PARAM_ERROR))
            return
        }

        val req = PayReq()
        req.appId = json.optString("appid")
        req.partnerId = json.optString("partnerid")
        req.prepayId = json.optString("prepayid")
        req.packageValue = json.optString("package")
        req.nonceStr = json.optString("noncestr")
        req.timeStamp = json.optString("timestamp")
        req.sign = json.optString("sign")
        wxApi.sendReq(req)
    }

    //支付回调响应
    fun onResp(error_code: Int) {
        when (error_code) {
            0 -> //成功
                mListener?.onSuccess()
            -1 -> //错误
                mListener?.onError(SocialError(SocialError.CODE_PAY_ERROR))
            -2 -> //取消
                mListener?.onCancel()
        }
        mListener = null
    }

    //检测是否支持微信支付
    private fun check(): Boolean {
        return wxApi.isWXAppInstalled && wxApi.wxAppSupportAPI >= Build.PAY_SUPPORTED_SDK_INT
    }

    companion object {
        private var instance: WXPay? = null
        fun getInstance(): WXPay {
            if (instance == null) {
                throw IllegalAccessException("请先调用initWxApi初始化支付接口")
            }
            return instance!!
        }

        fun initWxApi(wxApi: IWXAPI) {
            if (instance == null) {
                instance = WXPay(wxApi)
            }
        }
    }
}
